File: JVM/kotlin/Esinxe.kt
package com.esinxe

const val GOLDEN_GAMMA: ULong = 0x9E3779B97F4A7C15UL
const val MAX_INT_VALUE: ULong = 1_000_000_000_000_000_000UL
const val FNV_OFFSET_BASIS: ULong = 0xCBF29CE484222325UL
const val FNV_PRIME: ULong = 0x100000001B3UL

private val v1Prefix = byteArrayOf(101, 115, 105, 110, 120, 101, 45, 118, 49, 0)

sealed class Key {
    data class I64(val value: Long) : Key()
    data class U64(val value: ULong) : Key()
    data class StringValue(val value: String) : Key()
    data class BytesValue(val value: ByteArray) : Key()

    companion object {
        fun i64(value: Long): Key = I64(value)
        fun u64(value: ULong): Key = U64(value)
        fun string(value: String): Key = StringValue(value)
        fun bytes(value: ByteArray): Key = BytesValue(value.copyOf())
    }
}

fun mix64(input: ULong): ULong {
    var value = input
    value = (value xor (value shr 30)) * 0xBF58476D1CE4E5B9UL
    value = (value xor (value shr 27)) * 0x94D049BB133111EBUL
    return value xor (value shr 31)
}

private fun bounded(initial: ULong, maxValue: ULong): ULong {
    require(maxValue > 0UL) { "maxValue must be positive" }
    val threshold = (0UL - maxValue) % maxValue
    var value = initial
    var nonce = 0UL
    while (value < threshold) {
        nonce += 1UL
        value = mix64(value + nonce * GOLDEN_GAMMA)
    }
    return value % maxValue
}

private fun fnvUpdate(initial: ULong, data: ByteArray): ULong {
    var hash = initial
    for (value in data) {
        hash = (hash xor value.toUByte().toULong()) * FNV_PRIME
    }
    return hash
}

private fun fnvByte(hash: ULong, value: Int): ULong =
    (hash xor value.toULong()) * FNV_PRIME

private fun fnvULong(initial: ULong, value: ULong): ULong {
    var hash = initial
    for (shift in 0 until 64 step 8) {
        hash = fnvByte(hash, ((value shr shift) and 0xffUL).toInt())
    }
    return hash
}

private fun hashKey(initial: ULong, key: Key): ULong {
    return when (key) {
        is Key.I64 -> fnvULong(fnvByte(initial, 0x01), key.value.toULong())
        is Key.U64 -> fnvULong(fnvByte(initial, 0x02), key.value)
        is Key.StringValue -> {
            val data = key.value.encodeToByteArray()
            fnvUpdate(fnvULong(fnvByte(initial, 0x03), data.size.toULong()), data)
        }
        is Key.BytesValue ->
            fnvUpdate(
                fnvULong(fnvByte(initial, 0x04), key.value.size.toULong()),
                key.value,
            )
    }
}

class EsinxeRandom(seed: ULong) {
    private var seed: ULong = seed
    private var index: ULong = 0UL
    private var key: ULong = seed

    private fun keyedRaw(keys: Array<out Key>, domain: String? = null): ULong {
        var hash = fnvUpdate(FNV_OFFSET_BASIS, v1Prefix)
        hash = fnvULong(hash, seed)
        if (domain != null) {
            val data = domain.encodeToByteArray()
            hash = fnvByte(hash, 0xF0)
            hash = fnvULong(hash, data.size.toULong())
            hash = fnvUpdate(hash, data)
        }
        for (component in keys) {
            hash = hashKey(hash, component)
        }
        return mix64(hash)
    }

    fun raw(vararg keys: Key): ULong = keyedRaw(keys)

    fun int(maxValue: ULong, vararg keys: Key): ULong =
        bounded(raw(*keys), maxValue)

    fun range(minValue: Long, maxValue: Long, vararg keys: Key): Long {
        require(maxValue > minValue) { "maxValue must be greater than minValue" }
        val width = maxValue.toULong() - minValue.toULong()
        return (minValue.toULong() + bounded(raw(*keys), width)).toLong()
    }

    fun float01(vararg keys: Key): Double =
        (raw(*keys) shr 11).toDouble() / 9_007_199_254_740_992.0

    fun at2D(x: Long, y: Long, namespace: String? = null): ULong {
        val keys = mutableListOf<Key>(Key.i64(x), Key.i64(y))
        if (namespace != null) keys.add(Key.string(namespace))
        return keyedRaw(keys.toTypedArray(), "at2d")
    }

    fun at3D(x: Long, y: Long, z: Long, namespace: String? = null): ULong {
        val keys = mutableListOf<Key>(Key.i64(x), Key.i64(y), Key.i64(z))
        if (namespace != null) keys.add(Key.string(namespace))
        return keyedRaw(keys.toTypedArray(), "at3d")
    }

    fun chanceRatio(numerator: Long, denominator: Long, vararg keys: Key): Boolean {
        require(denominator > 0) { "denominator must be positive" }
        if (numerator <= 0) return false
        if (numerator >= denominator) return true
        return int(denominator.toULong(), *keys) < numerator.toULong()
    }

    fun <T> choose(items: List<T>, vararg keys: Key): T {
        require(items.isNotEmpty()) { "items must not be empty" }
        return items[int(items.size.toULong(), *keys).toInt()]
    }

    fun <T> shuffle(items: List<T>, vararg keys: Key): List<T> {
        val values = items.toMutableList()
        for (position in values.lastIndex downTo 1) {
            val iterationKeys = keys.toMutableList()
            iterationKeys.add(Key.u64(position.toULong()))
            val picked = bounded(
                keyedRaw(iterationKeys.toTypedArray(), "shuffle"),
                (position + 1).toULong(),
            ).toInt()
            val value = values[position]
            values[position] = values[picked]
            values[picked] = value
        }
        return values
    }

    fun <T> weightedChoice(
        items: List<T>,
        integerWeights: List<ULong>,
        vararg keys: Key,
    ): T {
        require(items.isNotEmpty() && items.size == integerWeights.size) {
            "items and weights must have the same non-zero length"
        }
        var total = 0UL
        for (weight in integerWeights) {
            val next = total + weight
            require(next >= total) { "total weight exceeds uint64" }
            total = next
        }
        require(total > 0UL) { "total weight must be positive" }
        val target = int(total, *keys)
        var running = 0UL
        for (position in items.indices) {
            running += integerWeights[position]
            if (target < running) return items[position]
        }
        error("unreachable weighted choice")
    }

    fun setSeed(seed: ULong) {
        this.seed = seed
        index = 0UL
        key = seed
    }

    fun nextRawAt(offset: ULong): ULong = mix64(seed + offset * GOLDEN_GAMMA)

    fun nextRaw(): ULong {
        val value = mix64(key)
        key += GOLDEN_GAMMA
        index += 1UL
        return value
    }

    fun nextAt(offset: ULong): ULong = bounded(nextRawAt(offset), MAX_INT_VALUE)

    fun next(): ULong {
        val value = bounded(mix64(key), MAX_INT_VALUE)
        key += GOLDEN_GAMMA
        index += 1UL
        return value
    }

    fun nextMaxAt(offset: ULong, maxValue: ULong): ULong =
        bounded(nextRawAt(offset), maxValue)

    fun nextMax(maxValue: ULong): ULong {
        val value = bounded(mix64(key), maxValue)
        key += GOLDEN_GAMMA
        index += 1UL
        return value
    }

    fun nextMinMaxAt(offset: ULong, minValue: ULong, maxValue: ULong): ULong? {
        if (maxValue <= minValue) return null
        return minValue + nextMaxAt(offset, maxValue - minValue)
    }

    fun nextMinMax(minValue: ULong, maxValue: ULong): ULong? {
        val value =
            if (maxValue <= minValue) null
            else minValue + bounded(mix64(key), maxValue - minValue)
        key += GOLDEN_GAMMA
        index += 1UL
        return value
    }
}

File: Go/esinxe/esinxe.go
package esinxe

import "errors"

const (
	GoldenGamma    = uint64(0x9E3779B97F4A7C15)
	MaxIntValue    = uint64(1000000000000000000)
	FNVOffsetBasis = uint64(0xCBF29CE484222325)
	FNVPrime       = uint64(0x100000001B3)
)

var v1Prefix = []byte("esinxe-v1\x00")

type KeyKind byte

const (
	SignedKey KeyKind = iota + 1
	UnsignedKey
	StringKey
	BytesKey
)

type Key struct {
	kind  KeyKind
	value uint64
	data  []byte
}

func I64(value int64) Key {
	return Key{kind: SignedKey, value: uint64(value)}
}

func U64(value uint64) Key {
	return Key{kind: UnsignedKey, value: value}
}

func String(value string) Key {
	return Key{kind: StringKey, data: []byte(value)}
}

func Bytes(value []byte) Key {
	copyValue := append([]byte(nil), value...)
	return Key{kind: BytesKey, data: copyValue}
}

func Mix64(value uint64) uint64 {
	value = (value ^ (value >> 30)) * 0xBF58476D1CE4E5B9
	value = (value ^ (value >> 27)) * 0x94D049BB133111EB
	return value ^ (value >> 31)
}

func bounded(value uint64, maxValue uint64) uint64 {
	if maxValue == 0 {
		return 0
	}
	threshold := -maxValue % maxValue
	nonce := uint64(0)
	for value < threshold {
		nonce++
		value = Mix64(value + nonce*GoldenGamma)
	}
	return value % maxValue
}

func fnvUpdate(hash uint64, data []byte) uint64 {
	for _, value := range data {
		hash ^= uint64(value)
		hash *= FNVPrime
	}
	return hash
}

func littleEndian64(value uint64) [8]byte {
	return [8]byte{
		byte(value),
		byte(value >> 8),
		byte(value >> 16),
		byte(value >> 24),
		byte(value >> 32),
		byte(value >> 40),
		byte(value >> 48),
		byte(value >> 56),
	}
}

func hashLength(hash uint64, length int) uint64 {
	encoded := littleEndian64(uint64(length))
	return fnvUpdate(hash, encoded[:])
}

func hashKey(hash uint64, key Key) uint64 {
	hash = fnvUpdate(hash, []byte{byte(key.kind)})
	if key.kind == SignedKey || key.kind == UnsignedKey {
		encoded := littleEndian64(key.value)
		return fnvUpdate(hash, encoded[:])
	}
	hash = hashLength(hash, len(key.data))
	return fnvUpdate(hash, key.data)
}

func keyedRaw(seed uint64, keys []Key, domain string) uint64 {
	hash := fnvUpdate(FNVOffsetBasis, v1Prefix)
	encodedSeed := littleEndian64(seed)
	hash = fnvUpdate(hash, encodedSeed[:])
	if domain != "" {
		hash = fnvUpdate(hash, []byte{0xF0})
		hash = hashLength(hash, len(domain))
		hash = fnvUpdate(hash, []byte(domain))
	}
	for _, key := range keys {
		hash = hashKey(hash, key)
	}
	return Mix64(hash)
}

type Random struct {
	seed  uint64
	index uint64
	key   uint64
}

func New(seed uint64) *Random {
	return &Random{seed: seed, key: seed}
}

func (r *Random) SetSeed(seed uint64) {
	r.seed = seed
	r.index = 0
	r.key = seed
}

func (r *Random) NextRawAt(offset uint64) uint64 {
	return Mix64(r.seed + offset*GoldenGamma)
}

func (r *Random) NextRaw() uint64 {
	value := Mix64(r.key)
	r.key += GoldenGamma
	r.index++
	return value
}

func (r *Random) Raw(keys ...Key) uint64 {
	return keyedRaw(r.seed, keys, "")
}

func (r *Random) Int(maxValue uint64, keys ...Key) (uint64, error) {
	if maxValue == 0 {
		return 0, errors.New("maxValue must be positive")
	}
	return bounded(r.Raw(keys...), maxValue), nil
}

func (r *Random) Range(minValue int64, maxValue int64, keys ...Key) (int64, error) {
	if maxValue <= minValue {
		return 0, errors.New("maxValue must be greater than minValue")
	}
	width := uint64(maxValue) - uint64(minValue)
	offset := bounded(r.Raw(keys...), width)
	return int64(uint64(minValue) + offset), nil
}

func (r *Random) Float01(keys ...Key) float64 {
	return float64(r.Raw(keys...)>>11) / float64(uint64(1)<<53)
}

func (r *Random) At2D(x int64, y int64, namespace *string) uint64 {
	keys := []Key{I64(x), I64(y)}
	if namespace != nil {
		keys = append(keys, String(*namespace))
	}
	return keyedRaw(r.seed, keys, "at2d")
}

func (r *Random) At3D(x int64, y int64, z int64, namespace *string) uint64 {
	keys := []Key{I64(x), I64(y), I64(z)}
	if namespace != nil {
		keys = append(keys, String(*namespace))
	}
	return keyedRaw(r.seed, keys, "at3d")
}

func (r *Random) ChanceRatio(numerator uint64, denominator uint64, keys ...Key) (bool, error) {
	if denominator == 0 {
		return false, errors.New("denominator must be positive")
	}
	if numerator >= denominator {
		return true, nil
	}
	return bounded(r.Raw(keys...), denominator) < numerator, nil
}

func Choose[T any](r *Random, items []T, keys ...Key) (T, error) {
	var zero T
	if len(items) == 0 {
		return zero, errors.New("items must not be empty")
	}
	index := bounded(r.Raw(keys...), uint64(len(items)))
	return items[index], nil
}

func Shuffle[T any](r *Random, items []T, keys ...Key) []T {
	values := append([]T(nil), items...)
	for index := len(values) - 1; index > 0; index-- {
		iterationKeys := append(append([]Key(nil), keys...), U64(uint64(index)))
		picked := bounded(
			keyedRaw(r.seed, iterationKeys, "shuffle"),
			uint64(index+1),
		)
		values[index], values[picked] = values[picked], values[index]
	}
	return values
}

func WeightedChoice[T any](r *Random, items []T, weights []uint64, keys ...Key) (T, error) {
	var zero T
	if len(items) == 0 || len(items) != len(weights) {
		return zero, errors.New("items and weights must have the same non-zero length")
	}
	total := uint64(0)
	for _, weight := range weights {
		next := total + weight
		if next < total {
			return zero, errors.New("total weight overflows uint64")
		}
		total = next
	}
	if total == 0 {
		return zero, errors.New("total weight must be positive")
	}
	target := bounded(r.Raw(keys...), total)
	running := uint64(0)
	for index, weight := range weights {
		running += weight
		if target < running {
			return items[index], nil
		}
	}
	return zero, errors.New("unreachable weighted choice")
}

func (r *Random) NextAt(offset uint64) uint64 {
	return bounded(r.NextRawAt(offset), MaxIntValue)
}

func (r *Random) Next() uint64 {
	value := bounded(Mix64(r.key), MaxIntValue)
	r.key += GoldenGamma
	r.index++
	return value
}

func (r *Random) NextMaxAt(offset uint64, maxValue uint64) uint64 {
	return bounded(r.NextRawAt(offset), maxValue)
}

func (r *Random) NextMax(maxValue uint64) uint64 {
	value := bounded(Mix64(r.key), maxValue)
	r.key += GoldenGamma
	r.index++
	return value
}

func (r *Random) NextMinMaxAt(offset uint64, minValue uint64, maxValue uint64) (uint64, bool) {
	if maxValue <= minValue {
		return minValue, false
	}
	return minValue + r.NextMaxAt(offset, maxValue-minValue), true
}

func (r *Random) NextMinMax(minValue uint64, maxValue uint64) (uint64, bool) {
	value := minValue
	ok := false
	if maxValue > minValue {
		value = minValue + bounded(Mix64(r.key), maxValue-minValue)
		ok = true
	}
	r.key += GoldenGamma
	r.index++
	return value, ok
}
